[mdoc] Adds custom formatting for the `ObjCRuntime.Platform` enum.

joelmartinez opened the pull request:
This enum, which can be found in Xamarin's iOS and Mac products use a custom encoding
scheme which resulted in invalid values being shown when processed with mdoc. This patch
adds a special case to watch for this particular enumeration value, and decodes the data
correctly.

Additionally, the code that formats the values for flags enumerations was refactored. Future extensibility of this feature can now be done by simply implementing a new `FlagsFormatter`, and applying the proper logic in the `MakeAttributesValueString` function.

joelmartinez commented:
Not sure what's up with the build failure ... as far as I could tell it wasn't due to any failure based on changes in this changeset (errors fetching git submodules in the build?)

jonpryor commented on mcs/tools/mdoc/Mono.Documentation/monodocer.cs:
```diff
@@ -5662,4 +5666,85 @@ protected override StringBuilder AppendNamespace (StringBuilder buf, TypeReferen
 	}
 }
 
+internal abstract class FlagsFormatter {
```
`internal` isn't needed.

jonpryor commented on mcs/tools/mdoc/Mono.Documentation/monodocer.cs:
```diff
@@ -5662,4 +5666,85 @@ protected override StringBuilder AppendNamespace (StringBuilder buf, TypeReferen
 }
 
+internal abstract class FlagsFormatter {
+	public abstract string Format(long c, Dictionary<long, string> values, string typename);
```
Space before `(`. This should likewise _probably_ be `IDictionary`, not `Dictionary`, and why `long`?

jonpryor replied:
What I think might make more sense is to:

1. Make `FlagsFormatter` a "normal" (non-`abstract`) class with a `virtual` `Format()` method.
2. `Format()` is `public virtual string Format(object value, TypeReference valueType)`

Then, _most_ of `MakeAttributesValueString()` becomes the base method implementation, and then it can be specialized:

    partial class FlagsFormatter {
        public string Format (object value, TypeReference valueType)
        {
            // existing MakeAttributesValueString() body, lines 2350-2365
            return Format (value, valueType.Resolve ());
        }

        protected virtual string Format (object value, TypeDefinition valueDef)
        {
            // existing MakeAttributesValueString() body lines 2365-end 
            // _except_ the enum-handling code on lines 2368 ... `return "(" ...`.
        }
    }

    class StandardFlagsEnumFormatter : FlagsFormatter {
        protected override string Format (object value, TypeDefinition valueDef)
        {
            if (valueDef.CustomAttributes.Any (ca => ca.AttributeType.FullName == "System.FlagsAttribute")) {
                return string.Join (" | ",
                        (from i in values.Keys
                         where (c & i) != 0
                         select typename + "." + values [i])
                        .DefaultIfEmpty (v.ToString ()).ToArray ());
            }
            return base.Format (value, valueDef);
    }

jonpryor replied:
The advantage to the above approach is that we don't "bake in" underlying enum types into the API, which makes things saner if/when we allow external add-ins -- we want as few assumptions/dependencies as possible. Here, all we rely on is the ability to get a Cecil TypeDefinition, as opposed to "all enum values are able to be stored in `long`s".

joelmartinez replied:
hmm, really interesting approach ... I only went with `long` because that's what the existing implementation was using for the flags enumeration formatting. I'll take a look at refactoring this part

jonpryor commented on mcs/tools/mdoc/Mono.Documentation/monodocer.cs:
```diff
@@ -5662,4 +5666,85 @@ protected override StringBuilder AppendNamespace (StringBuilder buf, TypeReferen
+		return string.Format("(Platform){0}", c);
+	}
+
+	string FormatValues (string plat, int arch, int major, int minor, int subminor) {
```
`{` on newline for methods, constructors.

jonpryor commented:
No expected HTML output? How do we ensure that the values/rendering are correct?

joelmartinez commented:
This piece isn't doing any of the HTML rendering ... it's simply formatting the value that goes into `<AttributeName>` ... this value is then *later* rendered to HTML as-is via monodoc.

joelmartinez commented:
@jonpryor ok ... I've submitted a now-refactored implementation. It is now much more amenable to a plug-in architecture in the future. Please let me know how it looks :)

jonpryor commented on mcs/tools/mdoc/Mono.Documentation/monodocer.cs:
```diff
@@ -2346,41 +2346,29 @@ private void MakeAttributes (XmlElement root, IEnumerable<string> attributes, Ty
-					return "typeof(" + NativeTypeManager.GetTranslatedName (vTypeRef) + ")"; // TODO: drop NS handling
-				else
-					return "typeof(" + v.ToString () + ")";
+		{
```
What happened to the indentation here? `MakeAttributesValueString()` is at 1 tab stop, yet this `{` is at 2 tab stops, and the body is at 3 tab stops?

I think Xamarin Studio is mucking things up...

joelmartinez replied:
ack! man, yeah Xamarin Studio makes it so hard to adhere to the existing formatting in this file ... I will submit another change that fixes this, but will wait until you're done looking at the rest of the changes :)

jonpryor commented on mcs/tools/mdoc/Mono.Documentation/monodocer.cs:
```diff
@@ -2346,41 +2346,29 @@ private void MakeAttributes (XmlElement root, IEnumerable<string> attributes, Ty
+				new AttributeValueFormatter (), 
+				new ApplePlatformEnumFormatter (), 
+				new StandardFlagsEnumFormatter (), 
+				new DefaultAttributeValueFormatter () 
```
Whenever possible (e.g. arrays, dictionaries), use a trailing `,`. This will allow you to append new values without needing to append a `,` to the previously-last value.

jonpryor commented on mcs/tools/mdoc/Mono.Documentation/monodocer.cs:
```diff
@@ -5662,4 +5650,171 @@ protected override StringBuilder AppendNamespace (StringBuilder buf, TypeReferen
+			string typename = MDocUpdater.GetDocTypeFullName (valueType);
+			var values = MDocUpdater.GetEnumerationValues (valueDef);
+			long c = MDocUpdater.ToInt64 (v);
+			if (c == 0 && values.Keys.Contains (c)) {
```
This case *should* have been handled in the `AttributeValueFormatter` type. Perhaps we should assert if we land here?

joelmartinez replied:
well, the reason I broke it out is because I had to place the logic for the "Apple" platform first, before doing this ... because technically, it would have been handled by this logic if it was a part of the first formatter in the array.

jonpryor replied:
You misunderstood;

> it would have been handled by this logic if it was a part of the first formatter in the array.

That's exactly the point. We have a condition which is handled in two separate places. Consequently, when reviewing this code *N* years from now, we'll be asking ourselves...are both of these code paths executed? Important? What happens if (when) they diverge?

An *assertion* would tell future maintainers that yes, this condition is duplicative, but it's not handled *here*, it's handled *there*, and if we get *here*, something is *wrong*:

    if (c == 0 && values.Keys.Contains (c)) {
        throw new InvalidOperationException ("Should not be reached; this circumstance should have been handled in AttributeValueFormatter.TryFormatValue()!");
    }

jonpryor commented on mcs/tools/mdoc/Mono.Documentation/monodocer.cs:
```diff
@@ -5662,4 +5650,171 @@ protected override StringBuilder AppendNamespace (StringBuilder buf, TypeReferen
+				returnvalue = string.Join (" | ",
+					(from i in values.Keys
+					where (c & i) == i && i != 0
+					select typename + "." + values [i])
```
Indentation: prefix `where`, `select` with a space:

    (from ...
     where ...
     select ... )

Alternatively, indent `where`, `select` by another tab level.

The intent is to not "hide" the parenthesis. Having `where` at the same indentation level as `(from` makes them look like peers (unless you read closely), e.g. for them to be peers it *should* be:

    (from  ... )
    where ...

(I can't think of any way for the above to be possible, but regardless...)

jonpryor commented on mcs/tools/mdoc/Test/DocTest-enumerations.cs:
```diff
@@ -0,0 +1,79 @@
+	[Flags]
+	public enum Platform : ulong {
+		None = 0,
+		iOS_2_0 = 131072,
```
Please use *hex* values here, to make it easier to figure out how things line up, bitmask-wise.

joelmartinez commented:
@jonpryor, made the requested changes :)

- test enum values now expressed as hex
- formatting of linq statement
- trailing `,` on array declaration
- indent level in `MakeAttributesValueString` function

The only feedback item that I didn't make any changes for was your comment on whether the functionality of `StandardFlagsEnumFormatter` should have been handled in `AttributeValueFormatter`. Please refer to the comment I made there as to why it's broken out into a separate class. Thanks!

jonpryor commented on mcs/tools/mdoc/Mono.Documentation/monodocer.cs:
```diff
@@ -5662,4 +5648,171 @@ protected override StringBuilder AppendNamespace (StringBuilder buf, TypeReferen
+class StandardFlagsEnumFormatter : AttributeValueFormatter {
+	public override bool TryFormatValue (object v, TypeReference valueType, out string returnvalue)
+	{
+		TypeDefinition valueDef = valueType.Resolve ();
```
My one last quibble is that `valueType.Resolve()` is potentially called multiple times. I'm not sure if that's a problem; is it cached?

(This is something that my previously suggested "inherit plus  delegate to `FormatValue(object, TypeDefinition)`" approach handled.)

joelmartinez commented:
@jonpryor 

- I now see what you were referring to. I actually only added that if after an earlier refactor, but you're totally right, and I removed that condition as it was unnecessary.
- I've also refactored to avoid calling `.Resolve ()` more than once. I took a peek in the Mono.Cecil code, and it doesn't look like it's cached, so I just made a small wrapper type to pass in rather than the bare `TypeReference`. The new `TypeInfo` has a property called `Definition` that will cache the result after the first call.

jonpryor commented:
> The new `TypeInfo` has a property

I would suggest *not* calling it `TypeInfo`, because `TypeInfo` is also the BCL type [System.Reflection.TypeInfo](https://msdn.microsoft.com/en-us/library/system.reflection.typeinfo(v=vs.110).aspx).

Perhaps `ResolvedTypeInfo`?

joelmartinez commented:
good point, type changed to `ResolvedTypeInfo`
